dbt/adapters/dremio/api/parameters.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (C) 2022 Dremio Corporation
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Optional
+from dbt.adapters.dremio.api.authentication import DremioAuthentication
+from dbt.adapters.dremio.credentials import DremioCredentials
+
+
+@dataclass
+class Parameters:
+    base_url: str
+    authentication: DremioAuthentication
+
+
+@dataclass
+class CloudParameters(Parameters):
+    cloud_project_id: str
+
+
+@dataclass
+class SoftwareParameters(Parameters):
+    pass
+
+
+@dataclass
+class ParametersBuilder:
+    authentication: Optional[DremioCredentials] = None
+
+    @classmethod
+    def __buildDremioAuthentication(self, credentials: DremioCredentials):
+        return DremioAuthentication.build(
+            credentials.UID, credentials.PWD, credentials.pat
+        )
+
+    @classmethod
+    def build(cls, credentials: DremioCredentials):
+        if credentials.cloud_host != None and credentials.cloud_project_id != None:
+            return CloudParametersBuilder(
+                cls.__buildDremioAuthentication(credentials=credentials),
+                credentials.cloud_host,
+                credentials.cloud_project_id,
+            )
+        elif (
+            credentials.software_host != None
+            and credentials.port != None
+            and credentials.use_ssl != None
```

Comment:
Isn't use_ssl populated by default for both cloud and software?

Reply:
Same with port

Reply:
I think checking the three parameters to be used is good defensive programming. Given that all of the variables are absolutely necessary in the UrlBuilder, and that users can change profiles.yml, I think it's good practice to ensure that they are all set.

Port also does not always default to 9047.[ Server admins can change ports in dremio.conf](https://docs.dremio.com/software/advanced-administration/dremio-conf/).

Reply:
In our dbt adapter, port and use_ssl will always default to a value (true and 9047, respectively), which means they will never be None, unless we make that change in DremioCredentials specifically.

Reply:
My point is that port will default to a value, doesn't matter if its 9047 or not

Reply:
Ah I see what you mean, it's defined in DremioCredentials and default to the values you've listed. I think I'd still prefer to keep the checks for legibility and just being defensive. If someone changes the DremioCredentials and no longer defaults them to a certain value, having this check would prevent a problem further down and lead to easier troubleshooting.

Reply:
We still need a case that checks for if someone provides a software_host and cloud_host. In this case, the software checks will pass and build software params.

Reply:
You make a good point. Let me make some changes.